[M2-5436][M2-5583] Add Participant > New Full/Limited Accounts

farmerpaul opened the pull request:
<!-- Use this template as a guide to describe your pull request, and adjust as necessary. -->
<!-- Include information that helps your peers review your updates and understand this    -->
<!-- repository's history of changes over time.                                           -->

<!-- Delete any options that are not relevant -->

Will complete before merging:
- [x] Tests for the changes have been added

### 📝 Description

<!-- Contributions are welcome! If there is a corresponding      -->
<!-- JIRA ticket, link to it by replacing `#` with ticket number -->

🔗 [Jira Ticket M2-5436](https://mindlogger.atlassian.net/browse/M2-5436)
🔗 [Jira Ticket M2-5583](https://mindlogger.atlassian.net/browse/M2-5583)

<!-- Replace this with a high-level description of the features/functionality proposed in the pull request. -->

This PR adds:

- Applet > Participants > **Add Participant** popup, with options to create new Full & Limited (Shell) accounts
- ~~"On-touch"~~ onBlur validation support for `InputController` & add tests
- Minor tweaks to modal styling to align with latest designs

> [!TIP]
> Changes have been organized by commit, each with an explanation of what's contained therein; it's recommended to review the PR [commit-by-commit](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1630/commits).

### 📸 Screenshots


https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/ce9a39a9-04d3-4e75-bfd0-f06cfc0462bb



### 🪤 Peer Testing

See Acceptance Criteria of Jira tickets linked above for detailed testing instructions.

farmerpaul committed:
M2-5436: Tweak Modal styling to align with designs

farmerpaul committed:
M2-5436: Add Large variant to ToggleButtons

farmerpaul committed:
M2-5436: Add validation features to form elements

Add ability to show error messages only after touch, and display
helperText on SelectController, if provided (aligns w/ InputController).

farmerpaul committed:
M2-5436: Create AddParticipantPopup

Creates 2-step popup, which utilizes new ToggleButtonGroup variant and
form validation UX.

Also update Spinner to use more reliable absolute positioning and sizing
(doesn't exceed container bounds when it has padding).

farmerpaul committed:
M2-5436: Create AddParticipantSuccessBanner

farmerpaul committed:
M2-5436: Update Participants screen

Integrates AddParticipantPopup into screen and updates column
Status => Account Type.

farmerpaul committed:
M2-5436: Update language strings

farmerpaul committed:
M2-5436: Fix broken tests

farmerpaul committed:
M2-5436: AddParticipantForm tests

farmerpaul committed:
M2-5436: simplify InputController + add tests

No need to tracking whether field was touched, which adds some
unintuitive logic to InputController. Added onBlur support instead,
which also provides enhanced UX.

farmerpaul committed:
M2-5436: add tests for AddParticipantPopup

sultanofcardio commented on src/modules/Dashboard/features/Participants/Participants.tsx:
```diff
@@ -518,6 +528,13 @@ export const Participants = () => {
         data-testid={`${dataTestid}-table`}
         {...tableProps}
       />
+      {appletId && addParticipantPopupVisible && (
+        <AddParticipantPopup
+          popupVisible={addParticipantPopupVisible}
+          appletId={appletId}
+          onClose={addParticipantOnClose}
+        />
+      )}
```
I think it might be useful if the `AddParticipantPopup` either accepted default data through props, or prevented dismissal on click outside (or both). It's currently easy to lose data via an accidental click.

I noticed that the acceptance criteria says navigation to the form after clicking the back button should clear it, and since re-opening the modal after clicking outside gives the overview it's kinda like clicking the back button, but I believe persisting the data unless the user explicitly closes the form with the X would be a better experience.

Something to consider.

farmerpaul replied:
Agreed! Good callout. Pushed a fix to support overriding of backdrop interactivity, and updated Modal tests as well.

sultanofcardio commented on src/modules/Dashboard/features/Applet/Popups/AddParticipantPopup/AddParticipantForm/AddParticipantForm.tsx:
```diff
@@ -0,0 +1,87 @@
+            <InputController
+              {...commonProps}
+              name={Fields.email}
+              label={t(Fields.email)}
```
The label here should read **Email Address**. There is already already an `emailAddress` resource value

https://github.com/ChildMindInstitute/mindlogger-admin/blob/7b0a8fb2f295363dbbe1b0efa5417bf5741c8b8c/src/resources/app-en.json#L1339

But the **address** is lowercase, so maybe you should create a new one (or uppercase it)

mbanting replied:
Good catch @sultanofcardio

sultanofcardio reviewed (approved):
Other than these two minor suggestions, looking good 👍

mbanting commented:
Very nice work @farmerpaul. Reviewed as well as tested most of the acceptance criteria, it all looks good.

farmerpaul committed:
M2-5436: ability to override modal backdrop click

As per suggested PR feedback, accidental clicking outside of modal
undesirably resets form, so override to prevent that is needed.
Also add Modal tests for backdrop click and left button click.

farmerpaul committed:
M2-5436: refine email field label

Also be consistent in how explicit we are in making use of i18n keys
so that it's easier to find out which keys the app is actually using.
(Obscuring them in an enum isn't very helpful.)
